rollup/eventwatcher/eventwatcher.go

```diff
@@ -0,0 +1,383 @@
+		scrollChainABI:                scrollChainABI,
+		l1CommitBatchEventSignature:   scrollChainABI.Events["CommitBatch"].ID,
+		l1RevertBatchEventSignature:   scrollChainABI.Events["RevertBatch"].ID,
+		l1FinalizeBatchEventSignature: scrollChainABI.Events["FinalizeBatch"].ID,
+		bc:                            bc,
+	}
+
+	return &service, nil
+}
+
+func (s *EventWatcher) Start() {
+	if s == nil {
+		return
+	}
+
+	log.Info("Starting batch event sync background service", "latest processed block", s.latestProcessedBlock)
+
+	go func() {
+		t := time.NewTicker(s.pollInterval)
+		defer t.Stop()
+
+		for {
+			s.fetchBatchEvents()
+
+			select {
+			case <-s.ctx.Done():
+				return
+			case <-t.C:
+				continue
+			}
+		}
+	}()
+}
+
+func (s *EventWatcher) Stop() {
+	if s == nil {
+		return
+	}
+
+	log.Info("Stopping batch event sync background service")
+
+	if s.cancel != nil {
+		s.cancel()
+	}
+}
+
+func (s *EventWatcher) fetchBatchEvents() {
+	latestConfirmed, err := s.client.getLatestFinalizedBlockNumber(s.ctx)
+	if err != nil {
+		log.Warn("failed to get latest confirmed block number", "err", err)
+		return
+	}
+
+	log.Trace("Sync service fetchBatchEvents", "latestProcessedBlock", s.latestProcessedBlock, "latestConfirmed", latestConfirmed)
+
+	// ticker for logging progress
+	t := time.NewTicker(logProgressInterval)
+
+	// query in batches
+	for from := s.latestProcessedBlock + 1; from <= latestConfirmed; from += defaultFetchBlockRange {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-t.C:
+			progress := 100 * float64(s.latestProcessedBlock) / float64(latestConfirmed)
+			log.Info("Syncing batch events", "processed", s.latestProcessedBlock, "confirmed", latestConfirmed, "progress(%)", progress)
+		default:
+		}
+
+		to := from + defaultFetchBlockRange - 1
+		if to > latestConfirmed {
+			to = latestConfirmed
+		}
+
+		logs, err := s.client.fetchBatchEventsInRange(s.ctx, from, to)
+		if err != nil {
+			// return and retry in next loop
+			log.Warn("failed to fetch batch events in range", "fromBlock", from, "toBlock", to, "err", err)
+			return
+		}
+
+		if err := s.parseAndUpdateBatchEventLogs(logs, to); err != nil {
+			// crash once update failed to avoid data inconsistency
+			log.Crit("failed to parse and update batch event logs", "err", err)
+		}
+	}
+}
+
+func (s *EventWatcher) parseAndUpdateBatchEventLogs(logs []types.Log, lastBlock uint64) error {
+	for _, vLog := range logs {
+		switch vLog.Topics[0] {
+		case s.l1CommitBatchEventSignature:
+			batchIndex := vLog.Topics[1].Big().Uint64()
+			chunkRanges, err := s.getChunkRanges(batchIndex, vLog)
+			if err != nil {
+				return fmt.Errorf("failed to get chunk ranges, err: %w", err)
+			}
+			rawdb.WriteBatchChunkRanges(s.db, batchIndex, chunkRanges)
+
+		case s.l1RevertBatchEventSignature:
+			batchIndex := vLog.Topics[1].Big().Uint64()
+			rawdb.DeleteBatchChunkRanges(s.db, batchIndex)
+
+		case s.l1FinalizeBatchEventSignature:
+			batchIndex := vLog.Topics[1].Big().Uint64()
```

Comment:
Use `UnpackLog(event, "FinalizeBatch", *log)` to get the `FinalizeBatch` event content.